File: web/src/main/java/net/danil/web/model/Problem.java
package net.danil.web.model;

import jakarta.persistence.*;
import lombok.*;

import java.util.List;

@Table
@Entity
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class Problem {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String name;
    // TODO: run OWASP Java HTML Sanitizer
    private String description;

    enum Difficulty {
        Easy, Normal, Hard
    }

    @Enumerated(EnumType.STRING)
    private Difficulty difficulty;

    @OneToMany(mappedBy = "problem", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<ProblemLanguage> languages;

    private Boolean active = false;
}

File: web/src/main/java/org/supremecode/web/domain/Problem.kt
package org.supremecode.web.domain

import jakarta.persistence.*
import java.util.LinkedHashSet

@Entity
@Table(name = "problem")
open class Problem() {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    open var id: Long? = null

    open var name: String = ""

    open var description: String = ""

    open var difficulty: String = ""

    @ManyToOne(fetch = FetchType.LAZY)
    open lateinit var author: User

    @OneToMany(fetch = FetchType.LAZY, mappedBy = "problem", cascade = [CascadeType.ALL], orphanRemoval = true)
    open var problemTags: MutableSet<ProblemTag> = LinkedHashSet()

    @OneToMany(mappedBy = "problem", cascade = [CascadeType.ALL], fetch = FetchType.LAZY, orphanRemoval = true)
    open var languages: MutableSet<ProblemLanguage> = LinkedHashSet()

    constructor(
        id: Long?,
        name: String,
        description: String,
        difficulty: String,
        languages: MutableSet<ProblemLanguage>,
        problemTags: MutableSet<ProblemTag>,
    ) : this() {
        this.id = id
        this.name = name
        this.description = description
        this.difficulty = difficulty
        this.languages = languages
        this.problemTags = problemTags
    }
}
